File: gcs_auth.py
#!/usr/bin/python
# Copyright (C) 2011 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# This code is not supported by Google

import sys
import getopt
import datetime
import time
import base64
import sys
import logging
import urllib,urllib2
from urllib2 import URLError, HTTPError
import json

from OpenSSL import crypto

"""
This script acquires an oauth2 bearer access_token use with authenticated GCS 
ShoppingAPI and ContentAPI requests.  GCS customers should use one of the 
libraries shown below to get the token in a production setting.  This script is 
provided to demonstrate the encryption and structure of getting an oauth2 bearer
 tokens. Ensure the computer this script is run on is synced with a NTP server.
 
 apt-get install python-openssl 
 
 install python crypto to
 /usr/local/lib/python2.6/dist-packages/Crypto

REFERENCE :
https://developers.google.com/commerce-search/docs/shopping_auth
https://developers.google.com/accounts/docs/OAuth2ServiceAccount
http://code.google.com/p/google-api-python-client/downloads/list
code.google.com/p/google-api-python-client/source/browse/oauth2client/crypt.py
http://packages.python.org/pyOpenSSL/
https://www.dlitz.net/software/pycrypto/

USAGE: 
    python gcs_auth.py --client_id= --key=

- arguments
  client_id: 'email' for the service account
  key: private key for the service account
  see
  https://developers.google.com/console/help/#service_accounts
  http://code.google.com/p/gcs-admin-toolkit/wiki/GCSAuthentication
  
- The output is an access_token that should be used in a GCS Search API call.
"""

class gcs_auth(object):
  
  SCOPES ='https://www.googleapis.com/auth/shoppingapi'
  SCOPES = SCOPES + ' https://www.googleapis.com/auth/structuredcontent'
  
  def __init__(self, client_id, key):
        
    try:
      f = file(key, 'rb')
      key = f.read()
    except IOError,e:
       self.log('Unable to open private key file: ' + key,logging.ERROR)
       sys.exit()
    f.close()
    pkey = crypto.load_pkcs12(key, 'notasecret').get_privatekey()
    
    jwt_header = '{"alg":"RS256","typ":"JWT"}'
    
    iss = client_id
    now = int(time.time())
    exptime = now + 3600
    claim =('{"iss":"%s",'
            '"scope":"%s",'
            '"aud":"https://accounts.google.com/o/oauth2/token",'
            '"exp":%s,'
            '"iat":%s}') %(iss,self.SCOPES,exptime,now)    

    jwt = self._urlsafe_b64encode(jwt_header) + '.' + \
          self._urlsafe_b64encode(unicode(claim, 'utf-8'))

    e = crypto.sign(pkey, jwt, 'sha256')
    assertion = jwt + '.' + self._urlsafe_b64encode(e)

    #self.log('header.claim: ' + jwt_header + '.' + claim, logging.DEBUG)
    #self.log('b64(header).b64(claim).b64(sign(b64(header).b64(claim))): ' + 
    #         assertion, logging.DEBUG)

    url = 'https://accounts.google.com/o/oauth2/token'
    data = {'grant_type' : 'assertion',
            'assertion_type' : 'http://oauth.net/grant_type/jwt/1.0/bearer',
            'assertion' : assertion }
    headers = {"Content-type": "application/x-www-form-urlencoded"}
     
    data = urllib.urlencode(data)
    req = urllib2.Request(url, data, headers)

    try:
      resp = urllib2.urlopen(req).read()
      parsed = json.loads(resp)
      self.access_token = parsed.get('access_token')
      self.log('access_token: ' + self.access_token,  logging.INFO)
    except HTTPError, e:
      self.log('Error code: ' + str(e.code),logging.ERROR)
      self.log(e.read(),logging.ERROR)
    except URLError, e:
      self.log( 'Reason: ' + str(e.reason),logging.ERROR)
      self.log(e.read(),logging.ERROR)      
      sys.exit(1)
      
  def read_token(self):
    return self.access_token
   
  # taken from /oauth2client/crypt.py   
  def _urlsafe_b64encode(self,raw_bytes):
    return base64.urlsafe_b64encode(raw_bytes).rstrip('=')
   
  def _urlsafe_b64decode(self,b64string):
    # Guard against unicode strings, which base64 can't handle.
    b64string = b64string.encode('ascii')
    padded = b64string + '=' * (4 - len(b64string) % 4)
    return base64.urlsafe_b64decode(padded)
      
#  http://code.google.com/p/google-api-python-client/
#  def googleAPIClient(self, client_id, key):
#    from oauth2client.client import SignedJwtAssertionCredentials
#    f = file(key, 'rb')
#    key = f.read()
#    f.close()
#    credentials = SignedJwtAssertionCredentials(
#        client_id,
#        key,
#        scope='https://www.googleapis.com/auth/shoppingapi')
#    http = httplib2.Http()
#    http = credentials.authorize(http)
#    credentials.refresh(http)
#    print credentials.access_token


  def log(self,msg, loglevel):
    #LOG_FILENAME = 'gcsoauth.log'
    #logging.basicConfig(filename=LOG_FILENAME,level=logging.INFO)
    m = ('[%s] %s') % (datetime.datetime.now(), msg)
    print m
    if (loglevel == logging.DEBUG):
      logging.debug(m)   
    else:
      logging.info(m)
               
if __name__ == '__main__':
  key = None
  client_id = None

  try:
    opts, args = getopt.getopt(sys.argv[1:], None, ["client_id=","key="])
  except getopt.GetoptError:
    print 'Please specify --client_id= --key='
    sys.exit(1)

  for opt, arg in opts:
    if opt == "--client_id":
      client_id = arg    
    if opt == "--key":
      key = arg
      
  if (key is not None and client_id is not None):
    gcs_auth(client_id,key)

File: gcs_auth.cc
#include <stdio.h>
#include <iostream>
#include <string.h>
#include <ctime>
#include <math.h>
#include <algorithm>

#include <sstream>
#include <stdexcept>
#include <iomanip>

#include <curl/curl.h>

#include <openssl/sha.h>  
#include <openssl/rsa.h>
#include <openssl/evp.h>
#include <openssl/objects.h>
#include <openssl/x509.h>
#include <openssl/err.h>
#include <openssl/pem.h>
#include <openssl/pkcs12.h> 
#include <openssl/bio.h>

using namespace std;

/*

apt-get install libcurl4-openssl-dev libssl-dev
g++  -std=c++11 -o gcs_auth gcs_auth.cc -lcurl -lcrypto 
./gcs_auth
{
  "access_token" : "ya29.3QDSgqs9QupWd2eqLdxrhu-vAVSdI35Ol2In61TymxxxxmC17pmf-G2y9KTqSTQPPFHGcOTsXEA",
  "token_type" : "Bearer",
  "expires_in" : 3600
}

This script acquires an oauth2 bearer access_token use with authenticated GCS 
ShoppingAPI and ContentAPI requests.  GCS customers should use one of the 
libraries shown below to get the token in a production setting.  This script 
is provided to  demonstrate the encryption and structure of getting an oauth2 
bearer tokens. Ensure the computer this script is run on is synced with a 
NTP server.

>> this code hans't been duration-tested in prod; may have a memory leak or two. <<

references/citations:

http://www.programmershare.com/1248592/
https://github.com/peervpn/peervpn/blob/master/libp2psec/rsa.c
http://sehermitage.web.fc2.com/program/src/rsacrypt.c
http://www.cplusplus.com/forum/unices/45878/
http://stackoverflow.com/questions/5288076/doing-base64-encoding-and-decoding-in-openssl-c
and others i thoughtlessly omitted

openssl pkcs12 -in xxxxyyyy.p12 -nocerts -out privateKey.pem
echo -n plainText | openssl dgst -sha256
(stdin)= 3ec8d98e737b84ff4fa0a9f0943d32bca35fcc73

*/

string data;

size_t writeCallback(char* buf, size_t size, size_t nmemb, void* up)
{ 
    for (int c = 0; c<size*nmemb; c++)
    {
        data.push_back(buf[c]);
    }
    return size*nmemb;
}

char *base64(const void *input, int length)
{
  BIO *bmem, *b64;
  BUF_MEM *bptr;

  b64 = BIO_new(BIO_f_base64());
  bmem = BIO_new(BIO_s_mem());
  b64 = BIO_push(b64, bmem);
  BIO_write(b64, input, length);
  BIO_flush(b64);
  BIO_get_mem_ptr(b64, &bptr);
  char *buff = (char *)malloc(bptr->length);
  memcpy(buff, bptr->data, bptr->length-1);

  buff[bptr->length-1] = 0;
  BIO_free_all(b64);
  return buff;
}

char * doSign(char*certFile,const char* pwd, string plainText)
{

    unsigned char hash[SHA256_DIGEST_LENGTH];
    unsigned char sign[256];
    unsigned int signLen;

    FILE* fp;
    if (!(fp = fopen(certFile, "rb"))) 
    { 
        fprintf(stderr, "Error opening file %s\n", certFile);        
        return NULL;     
    }    
    PKCS12 *p12= d2i_PKCS12_fp(fp, NULL);  
    fclose (fp);    
    if (!p12) {      
        fprintf(stderr, "Error reading PKCS#12 file\n");   
        ERR_print_errors_fp(stderr);  
        return NULL;   
    } 
     
    EVP_PKEY *pkey=NULL;     
    X509 *x509=NULL;
    STACK_OF(X509) *ca = NULL;
    if (!PKCS12_parse(p12, pwd, &pkey, &x509, &ca)) {         
        fprintf(stderr, "Error parsing PKCS#12 file\n");       
        ERR_print_errors_fp(stderr);
        return NULL;
    } 
    PKCS12_free(p12);

    int sigLen=EVP_PKEY_size(pkey);
    EVP_MD_CTX *ctx = EVP_MD_CTX_create();
    EVP_MD_CTX_init(ctx);

    RSA *prikey = EVP_PKEY_get1_RSA(pkey); 

   SHA256_CTX sha256;
   SHA256_Init(&sha256);
   const char * c = plainText.c_str();
   SHA256_Update(&sha256, c, strlen(c));
   SHA256_Final(hash, &sha256);

    stringstream ss;
    for(int i = 0; i < SHA256_DIGEST_LENGTH; i++)
    {
        ss << hex << setw(2) << setfill('0') << (int)hash[i];
    }
    //cout << "Text to sign ----->  " << plainText << "\n";
    //cout << "==================\n";    
    //cout << "SHA256:  " << ss.str() << "\n";
    //cout << "==================\n";

    int ret = RSA_sign(NID_sha256, hash, SHA256_DIGEST_LENGTH, sign,  &signLen, prikey);

    EVP_MD_CTX_destroy(ctx);
    RSA_free(prikey);
    EVP_PKEY_free(pkey);  
    X509_free(x509);

    return base64(sign,signLen);
}

int main(int argc, char* argv[]) {
    
    std::string SCOPE = "https://www.googleapis.com/auth/devstorage.read_write";
    char* certFile = (char*)"aaabbbbb.p12";
    std::string iss = "xxxxyyyyy@developer.gserviceaccount.com";

    std:string jwt_header = "{\"alg\":\"RS256\",\"typ\":\"JWT\"}";
    long now = std::time(0);
    std::string expire_on = std::to_string(now + 3600);
    string claim ="{\"iss\":\"" + iss + "\",\"scope\":\"" + SCOPE + "\",\"aud\":\"https://accounts.google.com/o/oauth2/token\",\"exp\":" + expire_on + ",\"iat\":" + std::to_string(now) + "}";
    //cout << "Clear text to sign ---> " << jwt_header << "." << claim << "\n";
    //cout << "==================\n";
    char* b64jwt = base64(jwt_header.c_str(), jwt_header.size());
    char* b64claim = base64(claim.c_str(), claim.size());    
    std::string jwt = std::string(b64jwt) + "." + std::string(b64claim);

    CRYPTO_malloc_init();
    ERR_load_crypto_strings();
    OpenSSL_add_all_algorithms(); 
    OpenSSL_add_all_ciphers();
    OpenSSL_add_all_digests();

    const char* passwd = "notasecret";
    char * e = doSign(certFile,passwd,jwt);  

    EVP_cleanup();
    //cout << "Signature: " << e << "\n";
    //cout << "jwt: =======\n" << jwt << "\n";
    std::string assertion = jwt + "." + e;
    //cout << "jwt + assertion ============\n" << assertion << "\n"; 

    replace(assertion.begin(),  assertion.end(),'+','-');
    replace(assertion.begin(),  assertion.end(),'/','_');
    //replace(assertion.begin(),  assertion.end(),'=','*');

    CURL* curl;

    curl_global_init(CURL_GLOBAL_ALL);
    curl=curl_easy_init();

    curl_easy_setopt(curl, CURLOPT_URL, "https://accounts.google.com/o/oauth2/token");
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &writeCallback);
    //curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L); 

    curl_easy_setopt(curl, CURLOPT_POST, 1);
    std::string  postfields =  std::string("grant_type=assertion&assertion_type=http%3A%2F%2Foauth.net%2Fgrant_type%2Fjwt%2F1.0%2Fbearer&assertion=") + assertion.c_str() ; 

    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postfields.c_str());    
    curl_easy_perform(curl);
    cout << endl << data << endl;
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    return 0;
}
